File: vision.cpp
//
// Created by Aditya Mangalampalli on 11/18/21.
//
#include <iostream>
#include <opencv2/opencv.hpp>
#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>

using namespace std;
using namespace cv;

//Mat &img = *(Mat *) frame_ptr;
int main() {
    VideoCapture camera(1);
    namedWindow("Webcam", WINDOW_AUTOSIZE);
    while (true) {
        Mat img;
        camera >> img;
        Mat hsv;
        GaussianBlur(img, hsv, Size(5, 5), BORDER_DEFAULT);
        cvtColor(hsv, hsv, COLOR_BGR2HSV);
        Mat masked;
        Scalar lower = Scalar(27, 26, 189);
        Scalar upper = Scalar(40, 255, 255);
        inRange(hsv, lower, upper, masked);
        bitwise_and(hsv, hsv, masked);
        cvtColor(hsv, hsv, COLOR_BGR2GRAY);
        Mat thresh_img;
        threshold(hsv, thresh_img, 100, 255, THRESH_BINARY + THRESH_OTSU);
        vector<vector<Point>> contour;
        vector<Vec4i> hierarchy;
        findContours(thresh_img, contour, hierarchy, RETR_EXTERNAL, CHAIN_APPROX_NONE);
        if (contour.empty()) {
            cout << 0 << endl;
            continue;
        } else {
            double max = 0;
            int index = 0;
            for (auto &i: contour) {
                if (contourArea(i) > max) {
                    max = contourArea(i);
                    index++;
                }
            }
            contour.erase(contour.begin(), contour.begin() + index - 1);
            contour.erase(contour.begin() + 1, contour.end());
        }
        int x = boundingRect(contour[0]).x;
        int y = boundingRect(contour[0]).y;
        int width = boundingRect(contour[0]).width;
        int height = boundingRect(contour[0]).height;
        int firstBound = int(img.size[1] / 3);
        int secondBound = int((img.size[1] / 3) * 2);
        rectangle(img, Point(x, y), Point((x + width), (y + height)), Scalar(0, 255, 0), 2);
        line(img, Point(firstBound, 0), Point(firstBound, img.size[0]), Scalar(255, 0, 0), 2);
        line(img, Point(secondBound, 0), Point(secondBound, img.size[0]), Scalar(255, 0, 0), 2);
        int centerX = int((x + (x + width)) / 2);
        if (centerX < firstBound) {
            putText(img, "Left", Point(x, y), FONT_HERSHEY_SIMPLEX, 1, (0, 0, 255), 2);
            cout << 1 << endl;
        } else if (centerX > secondBound) {
            putText(img, "Right", Point(x, y), FONT_HERSHEY_SIMPLEX, 1, (0, 0, 255), 2);
            cout << 2 << endl;
        } else {
            putText(img, "Center", Point(x, y), FONT_HERSHEY_SIMPLEX, 1, (0, 0, 255), 2);
            cout << 3 << endl;
        }
        imshow("Webcam", img);
        if (waitKey(10) == 27)
            break;
    }
    return 0;
}

File: python/vision.py
import cv2
import numpy as np

cap = cv2.VideoCapture(0)


def detectCube(frame):
    # Convert BGR to HSV
    hsv = cv2.GaussianBlur(frame, (5, 5), cv2.BORDER_DEFAULT)
    hsv = cv2.cvtColor(frame, cv2.COLOR_BGR2HSV)
    # define range of blue color in HSV
    lower = np.array([27, 26, 189])
    upper = np.array([40, 255, 255])
    masked = cv2.inRange(hsv, lower, upper)
    res = cv2.bitwise_and(frame, frame, mask=masked)
    ret, thresh_img = cv2.threshold(
        cv2.cvtColor(res, cv2.COLOR_BGR2GRAY),
        100,
        255,
        cv2.THRESH_OTSU + cv2.THRESH_BINARY,
    )
    contours, hierarchy = cv2.findContours(
        thresh_img, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_NONE
    )
    if len(contours) <= 0:
        return frame
    c = max(contours, key=cv2.contourArea)
    x, y, w, h = cv2.boundingRect(c)
    cv2.rectangle(frame, (x, y), (x + w, y + h), (0, 255, 0), 2)
    cv2.line(
        frame,
        (int(frame.shape[1] / 3), 0),
        (int((frame.shape[1] / 3)), frame.shape[0]),
        (255, 0, 0),
        2,
    )
    cv2.line(
        frame,
        (int((frame.shape[1] / 3) * 2), 0),
        (int((frame.shape[1] / 3) * 2), frame.shape[0]),
        (255, 0, 0),
        2,
    )
    centerX = int((x + (x + w)) / 2)
    if centerX < int((frame.shape[1] / 3)):
        cv2.putText(frame, "Left", (x, y), cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 0, 255), 2)
    elif centerX > int((frame.shape[1] / 3) * 2):
        cv2.putText(frame, "Right", (x, y), cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 0, 255), 2)
    else:
        cv2.putText(
            frame, "Center", (x, y), cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 0, 255), 2
        )
    return frame


while True:
    ret, frame = cap.read()
    cv2.imshow("frame", detectCube(frame))
    if cv2.waitKey(1) & 0xFF == ord("q"):
        break
cv2.destroyAllWindows()
